django/db/models/query.py

```diff
@@ -963,7 +963,26 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+
+            update_fields = set(defaults)
+            concrete_field_names = self.model._meta._concrete_field_names
+
+            # `update_fields` does not support non-concrete fields (#31382)
+            if concrete_field_names.issuperset(update_fields):
+
+                for field in self.model._meta.concrete_fields:
+                    # `auto_now` fields need to be within `update_fields` to be updated
+                    if (
+                        getattr(field, "auto_now", False)
+                        and field.name not in update_fields
+                    ):
+                        update_fields.add(field.name)
+                        if field.name != field.attname:
+                            update_fields.add(field.attname)
```

Comment:
Perhaps we could also move this out as a cached property alongside `_concrete_field_names`?

We'd probably want to name it in a way that can be used for any field that updates on save in case anything is added in the future, e.g. `_auto_update_field_names`.

Then this becomes:
```python
            update_fields = set(defaults)
            concrete_field_names = self.model._meta._concrete_field_names

            # `update_fields` does not support non-concrete fields (#31382)
            if concrete_field_names.issuperset(update_fields):
                # Add fields that are automatically updated, e.g. `auto_now` fields.
                # This is to maintain backward compatibility as these fields are not
                # updated if not specified explicitly in the `update_fields` list.
                update_fields |= self.model._meta._auto_update_field_names
                obj.save(using=self.db, update_fields=update_fields)

            else:
                obj.save(using=self.db)
```

Note that I've also updated the comment. We should be clear that, without this, things will break. Semi-related, I've [mentioned before](https://github.com/django/django/pull/13526#pullrequestreview-506449403) that it'd be nice to add some documentation that makes it clear that `update_fields` without specifying the fields with `add_now` won't update them.

Reply:
Pushed up a change thank you, looks cleaner :+1: 


Good point about adding to the docs, was looking [here](https://docs.djangoproject.com/en/4.1/ref/models/instances/#what-happens-when-you-save) where it says what happens when you save, point 2 says: 

> **Preprocess the data.** Each field’s [pre_save()](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.Field.pre_save) method is called to perform any automated data modification that’s needed. For example, the date/time fields override pre_save() to implement [auto_now_add](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.DateField.auto_now_add) and [auto_now](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.DateField.auto_now).

Which is maybe misleading in the `update_fields` case.
I can try to add a note in [here](https://docs.djangoproject.com/en/4.1/ref/models/instances/#what-happens-when-you-save) that this pre-processing step is skipped for those fields unless explicitly added into the `update_fields`.
Might not be the best place to put it/word it but I can give it a shot. @ngnpope what do you think?

Also should I put that in a separate PR or do you think it should be included in this one? :thinking:

Reply:
[e542e6f](https://github.com/django/django/pull/16072/commits/e542e6f12c7be90d17cf30258e362c33804d488c) I gave it a  go, let me know if it reads ok :+1: